File: language/src/main/java/cadenza/values/Closure.java
package cadenza.values;

import cadenza.types.Type;
import com.oracle.truffle.api.CompilerDirectives;
import com.oracle.truffle.api.RootCallTarget;
import com.oracle.truffle.api.frame.MaterializedFrame;
import com.oracle.truffle.api.interop.ArityException;
import com.oracle.truffle.api.interop.InteropLibrary;
import com.oracle.truffle.api.interop.TruffleObject;
import com.oracle.truffle.api.interop.UnsupportedTypeException;
import com.oracle.truffle.api.library.ExportLibrary;
import com.oracle.truffle.api.library.ExportMessage;
import com.oracle.truffle.api.nodes.ExplodeLoop;
import cadenza.nodes.*;

import java.util.Arrays;

@CompilerDirectives.ValueType
@ExportLibrary(InteropLibrary.class)
public class Closure implements TruffleObject {
  public final int arity; // local minimum arity to do anything, below this construct PAPs, above this pump arguments.
  public final Type type; // tells us the maximum number of arguments this can take and how to lint foreign arguments.

  public final RootCallTarget callTarget;
  public final MaterializedFrame env;

  // invariant: target should have been constructed from a FunctionBody
  // also assumes that env matches the shape expected by the function body
  public Closure(MaterializedFrame env, int arity, Type type, RootCallTarget callTarget) {
    assert callTarget.getRootNode() instanceof ClosureRootNode : "not a function body";
    assert (env != null) == ((ClosureRootNode)callTarget.getRootNode()).isSuperCombinator() : "calling convention mismatch";
    assert arity <= type.getArity();
    this.arity = arity;
    this.callTarget = callTarget;
    this.env = env;
    this.type = type;
  }

  // combinator
  public Closure(int arity, Type type, RootCallTarget callTarget) {
    this(null, arity, type, callTarget);
  }
  public final boolean isSuperCombinator() {
    return env != null;
  }

  @SuppressWarnings("SameReturnValue")
  @ExportMessage
  public final boolean isExecutable() { return true; }

  // allow the use of our closures from other polyglot languages
  @ExportMessage
  @ExplodeLoop
  public final Object execute(Object... arguments) throws ArityException, UnsupportedTypeException {
    int maxArity = type.getArity();
    int len = arguments.length;
    if (len > maxArity) throw ArityException.create(maxArity, len);
    Type currentType = type;
    for (Object argument : arguments) { // lint foreign arguments for safety
      Type.Arr arr = (Type.Arr) currentType; // safe by arity check
      arr.argument.validate(argument);
      currentType = arr.result;
    }
    return call(arguments);
  }

  // this logic needs to be copied into Code.App
  public final Object call(Object... arguments) {
    int len = arguments.length;
    if (len < arity) {
      return pap(this,arguments);
    } else if (len == arity) {
      return isSuperCombinator()
        ? callTarget.call(cons(env, arguments))
        : callTarget.call(arguments);
    } else {
      // the result _must_ be a closure.
      Closure next = (Closure)callTarget.call(consTake(env,arity,arguments));
      return next.call(drop(arity, arguments));
    }
  }

  // construct a partial application node, which should check that it is a PAP itself
  public Closure pap(Closure f, Object[] arguments) {
    return null; // TODO:
  }

  @ExplodeLoop
  private static Object[] cons(Object x, Object[] xs) {
    Object[] ys = new Object[xs.length + 1];
    ys[0] = x;
    System.arraycopy(xs, 0, ys, 1, xs.length);
    return ys;
  }
  @ExplodeLoop
  private static Object[] consTake(Object x, int n, Object[] xs) {
    Object[] ys = new Object[n + 1];
    ys[0] = x;
    System.arraycopy(xs, 0, ys, 1, n);
    return ys;
  }
  @ExplodeLoop
  private static Object[] drop(int k, Object[] xs) {
    return Arrays.copyOfRange(xs,k,xs.length);
  }

}

File: language/src/main/kotlin/cadenza/values/Closure.kt
package cadenza.values

import cadenza.nodes.*
import cadenza.types.*
import com.oracle.truffle.api.CompilerDirectives
import com.oracle.truffle.api.RootCallTarget
import com.oracle.truffle.api.frame.MaterializedFrame
import com.oracle.truffle.api.interop.ArityException
import com.oracle.truffle.api.interop.InteropLibrary
import com.oracle.truffle.api.interop.TruffleObject
import com.oracle.truffle.api.interop.UnsupportedTypeException
import com.oracle.truffle.api.library.ExportLibrary
import com.oracle.truffle.api.library.ExportMessage
import com.oracle.truffle.api.nodes.ExplodeLoop

@CompilerDirectives.ValueType
@ExportLibrary(InteropLibrary::class)
class Closure (
  private val env: MaterializedFrame? = null,
  val arity: Int,
  val type: Type,
  val callTarget: RootCallTarget
) : TruffleObject {
  @Suppress("NOTHING_TO_INLINE")
  private inline fun isSuperCombinator() = env != null

  init {
    assert(callTarget.rootNode is ClosureRootNode) { "not a function body" }
    assert(env != null == (callTarget.rootNode as ClosureRootNode).isSuperCombinator()) { "calling convention mismatch" }
    assert(arity <= type.arity)
  }

  @ExportMessage
  fun isExecutable() = true

  // allow the use of our closures from other polyglot languages
  @ExportMessage
  @ExplodeLoop
  @Throws(ArityException::class, UnsupportedTypeException::class)
  fun execute(vararg arguments: Any?): Any? {
    val maxArity = type.arity
    val len = arguments.size
    if (len > maxArity) throw ArityException.create(maxArity, len)
    var currentType = type
    for (argument in arguments) { // lint foreign arguments for safety
      val arr = currentType as Arr // safe by arity check
      arr.argument.validate(argument)
      currentType = arr.result
    }
    @Suppress("UNCHECKED_CAST")
    return call(arguments as Array<Any?>)
  }

  fun call(arguments: Array<Any?>): Any? {
    val len = arguments.size
    return when {
      len < arity -> pap(arguments)
      len == arity ->
        if (isSuperCombinator()) callTarget.call(cons(env, arguments))
        else callTarget.call(arguments)
      else -> (callTarget.call(consTake(env, arity, arguments)) as Closure).call(drop(arity, arguments))
    }
  }

  // construct a partial application node, which should check that it is a PAP itself
  @ExplodeLoop
  fun pap(@Suppress("UNUSED_PARAMETER") arguments: Array<Any?>): Closure? {
    return null // TODO
  }
}

// TODO: incompatible, pick one
@ExplodeLoop
@Suppress("NOTHING_TO_INLINE")
private inline fun <reified T> cons(x: T, xs: Array<T>): Array<T> {
  val ys = arrayOfNulls<T>(xs.size + 1)
  ys[0] = x
  System.arraycopy(xs, 0, ys, 1, xs.size)
  @Suppress("UNCHECKED_CAST")
  return ys as Array<T>
}

@ExplodeLoop
@Suppress("NOTHING_TO_INLINE")
private inline fun <reified T> consTake(x: T, n: Int, xs: Array<T>): Array<T> {
  val ys = arrayOfNulls<T>(n + 1)
  ys[0] = x
  System.arraycopy(xs, 0, ys, 1, n)
  @Suppress("UNCHECKED_CAST")
  return ys as Array<T>
}

@ExplodeLoop
@Suppress("NOTHING_TO_INLINE")
private inline fun <T> drop(k: Int, xs: Array<T>): Array<T> {
  return xs.copyOfRange(k, xs.size)
}